mountpoint-s3/src/inode.rs

```diff
@@ -354,17 +358,100 @@ impl Superblock {
+
+            WriteStatus::Remote => {
+                error!(parent = parent_ino, ?name, "rmdir called on a remote directory",);
+                return Err(InodeError::CannotRemoveRemoteDirectory(inode.ino()));
+            }
+
+            WriteStatus::LocalUnopened => {
+                // A local directory can only contain local children.
+                match &inode_state.kind_data {
+                    InodeKindData::File {} => unreachable!("Already checked that inode is a directory"),
+                    InodeKindData::Directory {
+                        writing_children,
+                        deleted,
+                        ..
+                    } => {
+                        if *deleted {
+                            return Err(InodeError::InodeDoesNotExist(inode.ino()));
+                        }
+                        if !writing_children.is_empty() {
+                            return Err(InodeError::DirectoryNotEmpty(inode.ino()));
+                        }
+                    }
+                }
+            }
+        }
+
+        match &mut parent_state.kind_data {
+            InodeKindData::File {} => {
+                debug_assert!(false, "inodes never change kind");
+                return Err(InodeError::NotADirectory(parent.ino()));
+            }
+            InodeKindData::Directory {
+                children,
+                writing_children,
+                ..
+            } => {
+                assert!(writing_children.remove(&inode.ino()));
+                children.remove(inode.name());
+
+                match &mut inode_state.kind_data {
+                    InodeKindData::File {} => unreachable!("Already checked that inode is a directory"),
+                    InodeKindData::Directory { deleted, .. } => *deleted = true,
+                }
+            }
+        }
+
+        Ok(())
+    }
 }
 
 impl SuperblockInner {
     /// Retrieve the inode for the given number if it exists
     pub fn get(&self, ino: InodeNo) -> Result<Inode, InodeError> {
-        self.inodes
-            .read()
-            .unwrap()
-            .get(&ino)
-            .cloned()
-            .ok_or(InodeError::InodeDoesNotExist(ino))
+        let inode = self.inodes.read().unwrap().get(&ino).cloned();
+
+        match inode {
+            Some(inode) => {
+                let inode_state = inode.inner.sync.read().unwrap();
+                match &inode_state.kind_data {
+                    // TODO: Include removal check for files
+                    InodeKindData::File {} => Ok(inode.clone()),
+                    InodeKindData::Directory { deleted, .. } => {
+                        if *deleted {
+                            Err(InodeError::InodeDoesNotExist(ino))
+                        } else {
+                            Ok(inode.clone())
+                        }
+                    }
+                }
+            }
+            None => Err(InodeError::InodeDoesNotExist(ino)),
+        }
```

Comment:
a little less indent-y with `?` and match-if:
```suggestion
        let inode = self
            .inodes
            .read()
            .unwrap()
            .get(&ino)
            .cloned()
            .ok_or(InodeError::InodeDoesNotExist(ino))?;
        let inode_state = inode.inner.sync.read().unwrap();
        match &inode_state.kind_data {
            // TODO: Include removal check for files
            InodeKindData::Directory { deleted, .. } if *deleted => Err(InodeError::InodeDoesNotExist(ino)),
            _ => Ok(inode.clone()),
        }
```

Reply:
Done